extend check opensearch_service_domains_use_cognito_authentication_for_kibana

kagahd opened the pull request:
### Context 

The check `opensearch_service_domains_use_cognito_authentication_for_kibana` fails if Amazon OpenSearch has Amazon Cognito authentication for Kibana **not** enabled. However, besides Cognito, AWS recommends to use **SAML** authentication for Kibana which means that Cognito must be disabled. Since the `opensearch_service_domains_use_cognito_authentication_for_kibana` does not test whether **either** Cognito **or** SAML is enabled, it will fail when SAML is enabled.


### Description

Therefore, I renamed the check `opensearch_service_domains_use_cognito_authentication_for_kibana` to `opensearch_service_domains_use_cognito_or_saml_authentication_for_kibana` and extended it in a way that it tests whether **either** Cognito **or** SAML is enabled for Kibana on AWS OpenSearch. If either is enabled, `opensearch_service_domains_use_cognito_or_saml_authentication_for_kibana` will pass.
That's by the way also what the corresponding prowler v2 check `check_extra780` did at its time. See its corresponding [pull request](https://github.com/prowler-cloud/prowler/pull/1291) that had been merged into main at that time. It's a pity that this change hadn't been migrated neither to prowler v3 nor to v4.

Since my company has not yet migrated from prowler v3 to v4, I'm opening this PR for v3 first. Do you want me to open a PR also for prowler v4?

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jfagoagas commented:
Hi @kagahd, thanks for your contribution!

We'll review the PR as soon as possible but some things to consider:
- We cannot rename the check since it'll introduce a breaking change in the outputs and unique ids.
- If you can open this PR too to v4 it'd be great.
- We have to include more tests cases to handle this new behaviour, I can help you with that.

kagahd commented:
Hi @jfagoagas , thanks for your fast reply.
Ok, I rename the check to its former name but then its title would be a bit misleading. Is this ok?
Great to have your help for adding more unit tests!

jfagoagas commented:
> Hi @jfagoagas , thanks for your fast reply. Ok, I rename the check to its former name but then its title would be a bit misleading. Is this ok? Great to have your help for adding more unit tests!

The title and description can be kept as you did.

kagahd commented:
I already renamed back the check to its original name before I installed `pre-commit`? Do I need to commit and push again to let `pre-commit` kick in?

jfagoagas commented:
> I already renamed back the check to its original name before I installed `pre-commit`? Do I need to commit and push again to let `pre-commit` kick in?

I've just made some changes for you to review it but I need push access to your forked branch, could you give me access?

kagahd commented:
> I've just made some changes for you to review it but I need push access to your forked branch, could you give me access?

I added you.
I also added unit tests. I hope we were not cross-working.

jfagoagas commented on tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py:
```diff
@@ -87,7 +88,39 @@ def test_cognito_enabled(self):
+
+            check = opensearch_service_domains_use_cognito_authentication_for_kibana()
+            result = check.execute()
+            assert len(result) == 1
+            assert result[0].status == "PASS"
+            assert search(
+                "has either Amazon Cognito authentication or SAML authentication for Kibana enabled",
                 result[0].status_extended,
             )
             assert result[0].resource_id == domain_name
```
```suggestion
            assert len(result) == 1
            assert result[0].status == "PASS"
            assert (
                result[0].status_extended
                == f"Opensearch domain {domain_name} has either Amazon Cognito or SAML authentication for Kibana enabled."
            )
            assert result[0].region == AWS_REGION_EU_WEST_1
            assert result[0].resource_id == domain_name
            assert result[0].resource_arn == domain_arn
            assert result[0].resource_tags == []
```

jfagoagas commented on tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py:
```diff
@@ -87,7 +88,39 @@ def test_cognito_enabled(self):
+                saml_enabled=True,
+            )
+        )
+        opensearch_client.opensearch_domains[0].logging = []
```
```suggestion
```

jfagoagas commented on tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py:
```diff
@@ -87,7 +88,39 @@ def test_cognito_enabled(self):
             assert len(result) == 1
```
Please change this test assert by
```
assert len(result) == 1
            assert result[0].status == "PASS"
            assert (
                result[0].status_extended
                == f"Opensearch domain {domain_name} has either Amazon Cognito or SAML authentication for Kibana enabled."
            )
            assert result[0].region == AWS_REGION_EU_WEST_1
            assert result[0].resource_id == domain_name
            assert result[0].resource_arn == domain_arn
            assert result[0].resource_tags == []
```

jfagoagas commented on tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py:
```diff
@@ -55,7 +56,7 @@ def test_no_cognito_enabled(self):
             assert result[0].status == "FAIL"
             assert search(
-                "does not have Amazon Cognito authentication for Kibana enabled",
+                "has neither Amazon Cognito authentication nor SAML authentication for Kibana enabled.",
```
The same here, change this test assert to
```
assert len(result) == 1
            assert result[0].status == "FAIL"
            assert (
                result[0].status_extended
                == f"Opensearch domain {domain_name} has neither Amazon Cognito nor SAML authentication for Kibana enabled."
            )
            assert result[0].region == AWS_REGION_EU_WEST_1
            assert result[0].resource_id == domain_name
            assert result[0].resource_arn == domain_arn
            assert result[0].resource_tags == []
```

jfagoagas commented on prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.py:
```diff
@@ -14,10 +14,10 @@ def execute(self):
             report.status = "PASS"
-            report.status_extended = f"Opensearch domain {domain.name} has Amazon Cognito authentication for Kibana enabled."
-            if not domain.cognito_options:
+            report.status_extended = f"Opensearch domain {domain.name} has either Amazon Cognito authentication or SAML authentication for Kibana enabled."
```
```suggestion
            report.status_extended = f"Opensearch domain {domain.name} has either Amazon Cognito or SAML authentication for Kibana enabled."
```

jfagoagas commented on prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.py:
```diff
@@ -14,10 +14,10 @@ def execute(self):
+            if not domain.cognito_options and not domain.saml_enabled:
                 report.status = "FAIL"
-                report.status_extended = f"Opensearch domain {domain.name} does not have Amazon Cognito authentication for Kibana enabled."
+                report.status_extended = f"Opensearch domain {domain.name} has neither Amazon Cognito authentication nor SAML authentication for Kibana enabled."
```
```suggestion
                report.status_extended = f"Opensearch domain {domain.name} has neither Amazon Cognito nor SAML authentication for Kibana enabled."
```

jfagoagas commented on prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.metadata.json:
```diff
@@ -11,8 +11,8 @@
   "ResourceType": "AwsOpenSearchDomain",
-  "Description": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito authentication for Kibana enabled",
-  "Risk": "Amazon Elasticsearch Service supports Amazon Cognito for Kibana authentication.",
+  "Description": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito authentication or SAML authentication for Kibana enabled",
```
```suggestion
  "Description": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito or SAML authentication for Kibana enabled",
```

jfagoagas commented on prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.metadata.json:
```diff
@@ -1,7 +1,7 @@
   "Provider": "aws",
   "CheckID": "opensearch_service_domains_use_cognito_authentication_for_kibana",
-  "CheckTitle": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito authentication for Kibana enabled",
+  "CheckTitle": "Check if Amazon Elasticsearch/Opensearch Service domains has either Amazon Cognito authentication or SAML authentication for Kibana enabled",
```
```suggestion
  "CheckTitle": "Check if Amazon Elasticsearch/Opensearch Service domains has either Amazon Cognito or SAML authentication for Kibana enabled",
```

jfagoagas commented on tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py:
```diff
@@ -38,6 +37,7 @@ def test_no_cognito_enabled(self):
+                saml_enabled=False,
             )
         )
         opensearch_client.opensearch_domains[0].logging = []
```
```suggestion
```

jfagoagas reviewed (approved):
That's a great improvement for the check @kagahd 👏 

Thanks for taking care of it also including new tests cases 🚀 

If you can include this work for the v4 it'd be really appreciated 🙌
